src/core/comm/coll.cc

```diff
@@ -0,0 +1,443 @@
+{
+  // IN_PLACE
+  if (sendbuf == recvbuf) { log_coll.fatal("Do not support inplace Alltoall"); }
+  log_coll.debug("Alltoall: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#ifdef LEGATE_USE_GASNET
+  return alltoallMPI(sendbuf, recvbuf, count, type, global_comm);
+#else
+  return alltoallLocal(sendbuf, recvbuf, count, type, global_comm);
+#endif
+}
+
+int collAllgather(
+  const void* sendbuf, void* recvbuf, int count, CollDataType type, CollComm global_comm)
+{
+  log_coll.debug("Allgather: global_rank %d, mpi_rank %d, unique_id %d, comm_size %d",
+                 global_comm->global_rank,
+                 global_comm->mpi_rank,
+                 global_comm->unique_id,
+                 global_comm->global_comm_size);
+#ifdef LEGATE_USE_GASNET
+  return allgatherMPI(sendbuf, recvbuf, count, type, global_comm);
+#else
+  return allgatherLocal(sendbuf, recvbuf, count, type, global_comm);
+#endif
+}
+
+// called from main thread
+int collInit(int argc, char* argv[])
+{
+  current_unique_id    = 0;
+  const char* nb_comms = getenv("LEGATE_MAX_CPU_COMMS");
+  if (nb_comms != nullptr) { MAX_NB_COMMS = atoi(nb_comms); }
+  assert(MAX_NB_COMMS > 0);
+#ifdef LEGATE_USE_GASNET
+  int provided, init_flag = 0;
+  CHECK_MPI(MPI_Initialized(&init_flag));
+  if (!init_flag) {
+    CHECK_MPI(MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided));
+  } else {
+    int mpi_thread_model;
+    MPI_Query_thread(&mpi_thread_model);
+    if (mpi_thread_model != MPI_THREAD_MULTIPLE) {
+      log_coll.fatal(
+        "MPI has been initialized by others, but is not sinitialized with "
+        "MPI_THREAD_MULTIPLE");
+    }
+  }
+  mpi_comms.resize(MAX_NB_COMMS, MPI_COMM_NULL);
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_dup(MPI_COMM_WORLD, &mpi_comms[i])); }
+#else
+  thread_comms.resize(MAX_NB_COMMS);
+  for (int i = 0; i < MAX_NB_COMMS; i++) {
+    thread_comms[i].ready_flag = false;
+    thread_comms[i].buffers    = nullptr;
+    thread_comms[i].displs     = nullptr;
+  }
+#endif
+  coll_inited = true;
+  return CollSuccess;
+}
+
+int collFinalize(void)
+{
+  assert(coll_inited == true);
+  coll_inited = false;
+#ifdef LEGATE_USE_GASNET
+  for (int i = 0; i < MAX_NB_COMMS; i++) { CHECK_MPI(MPI_Comm_free(&mpi_comms[i])); }
+  mpi_comms.clear();
+  return MPI_Finalize();
+#else
+  for (int i = 0; i < MAX_NB_COMMS; i++) { assert(thread_comms[i].ready_flag == false); }
+  thread_comms.clear();
+  return CollSuccess;
+#endif
+}
+
+int collGetUniqueId(int* id)
+{
+  *id = current_unique_id;
+  current_unique_id++;
+  if (current_unique_id > MAX_NB_COMMS) {
+    log_coll.fatal(
+      "Please increase the LEGATE_MAX_CPU_COMMS by export LEGATE_MAX_CPU_COMMS=new number, current "
```

Comment:
Chage `export LEGATE_MAX_CPU_COMMS=new number` to `\"export LEGATE_MAX_CPU_COMMS=(new number)\"`.

Reply:
fixed.